### PR TITLE
Add fee estimate method to output manager service and wallet_ffi

### PR DESCRIPTION
**Pull request by delta1**
## Description

Adds a `fee_estimate` method to the output manager service and also to the wallet_ffi 

## Motivation and Context

This will allow the mobile app to show accurate fees for the transaction 

## How Has This Been Tested?

- `cargo test -p tari_wallet`
- `cargo test -p tari_wallet_ffi`

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/wallet/src/output_manager_service/handle.rs [hunk 1]
```diff
@@ -131,6 +134,7 @@ pub enum OutputManagerEvent {
     TxoValidationSuccess(u64),
     TxoValidationFailure(u64),
     TxoValidationAborted(u64),
+    FeeEstimate(MicroTari),
```
Comment (philipr-za) [inline base_layer/wallet/src/output_manager_service/handle.rs]: I don't think this is the appropriate way to return this value. The event bus is used to return feedback for long running async tasks. These are typically operations that require communications to external nodes.

This operation only requires a single DB read to selection the outputs so I think this should be returned directly by the Output Manager Service api call.

File: base_layer/wallet/src/output_manager_service/handle.rs [hunk 2]
```diff
@@ -236,6 +240,31 @@ impl OutputManagerHandle {
...
 
+    /// Get a fee estimate for an amount of MicroTari, at a specified fee per gram and given number of kernels and
+    /// outputs
+    pub async fn fee_estimate(
+        &mut self,
+        amount: MicroTari,
+        fee_per_gram: MicroTari,
+        num_kernels: u64,
+        num_outputs: u64,
+    ) -> Result<MicroTari, OutputManagerError>
```
Comment (philipr-za) [inline base_layer/wallet/src/output_manager_service/handle.rs:250]: So here I see it is returned by API call in the handle so then I am not sure why you need to add a variant to the event enum.

File: base_layer/wallet/tests/output_manager_service/service.rs [hunk 3]
```diff
@@ -275,6 +275,51 @@ fn sending_transaction_and_confirmation<T: Clone + OutputManagerBackend + 'stati
...
+    // minimum fee
+    let fee_per_gram = MicroTari::from(1);
+    let fee = runtime
+        .block_on(oms.fee_estimate(MicroTari::from(100), fee_per_gram, 1, 1))
+        .unwrap();
+    assert_eq!(fee, MicroTari::from(100));
+
+    let fee_per_gram = MicroTari::from(25);
+    let fee = runtime
+        .block_on(oms.fee_estimate(MicroTari::from(100), fee_per_gram, 1, 1))
```
Comment (philipr-za) [inline base_layer/wallet/tests/output_manager_service/service.rs]: I think it's important to test this while varying the number of outputs in a transaction.

File: base_layer/wallet/src/output_manager_service/handle.rs (hunk 2 above)
Comment (delta1) [reply base_layer/wallet/src/output_manager_service/handle.rs:250]: My mistake

File: base_layer/wallet/tests/output_manager_service/service.rs (hunk 3 above)
Comment (delta1) [reply base_layer/wallet/tests/output_manager_service/service.rs]: will do

File: base_layer/wallet/src/output_manager_service/handle.rs (hunk 1 above)
Comment (delta1) [reply base_layer/wallet/src/output_manager_service/handle.rs]: sorry this was mistakenly included

**Review by philipr-za — APPROVED** [review]
🎉

File: base_layer/wallet_ffi/wallet.h [hunk 4]
```diff
@@ -461,6 +461,9 @@ unsigned long long wallet_get_pending_incoming_balance(struct TariWallet *wallet
 // Gets the outgoing balance from a TariWallet
 unsigned long long wallet_get_pending_outgoing_balance(struct TariWallet *wallet,int* error_out);
 
+// Get a fee estimate from a TariWallet for a given amount
+unsigned long long wallet_get_fee_estimate(struct TariWallet *wallet, unsigned long long amount, unsigned long long fee_per_gram, unsigned long long num_kernels, unsigned long long num_outputs, int* error_out);
```
Comment (StriderDM) [inline base_layer/wallet_ffi/wallet.h:465]: @delta1 When this merges, this method still needs to be added to the wrappers for iOS and Android. Just mentioning this here.

Comment (delta1) [reply base_layer/wallet_ffi/wallet.h:465]: thanks @StriderDM agreed

**Pull request merged**